src/perform_step/adams_bashforth_moulton_perform_step.jl

```diff
@@ -0,0 +1,123 @@
+function initialize!(integrator,cache::AB3ConstantCache)
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t) # Pre-start fsal
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+@muladd function perform_step!(integrator,cache::AB3ConstantCache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  k1 = integrator.fsalfirst
+  cnt = length(integrator.sol.t)
+  if cnt == 1 || cnt == 2
+    ttmp = t + (2/3)*dt
+    ralk2 = f(uprev + (2/3)*dt*k1, p, ttmp)
+    u = uprev + (dt/4)*(k1 + 3*ralk2)
+  else
+    u1 = integrator.sol(t-dt)
+    u2 = integrator.sol(t-2*dt)
+    k2 = f(u1, p, t-dt)
+    k3 = f(u2, p, t-2*dt)
```

Comment:
You don't need to calculate thi. `k3` is `k2` of last time. `k2` is `k1` of last time. And then `k1` comes from `fsalfirst` calculations. So you actually don't need any function calls other than the FSAL interpolation one.